crates/wast/src/core/types.rs

```diff
@@ -57,11 +57,57 @@ impl<'a> Peek for ValType<'a> {
     }
 }
 
-/// A heap type for a reference type
+/// A heap type for a reference type.
 #[allow(missing_docs)]
 #[derive(Debug, PartialEq, Eq, Hash, Copy, Clone)]
 pub enum HeapType<'a> {
-    /// An untyped function reference: funcref. This is part of the reference
+    /// An abstract reference. With the shared-everything-threads proposal,
+    /// these types can also be marked `shared`.
+    Abstract { shared: bool, ty: AbstractHeapType },
+    /// A reference to a concrete function, struct, or array type defined by
+    /// Wasm: `ref T`. This is part of the function references and GC proposals.
+    Concrete(Index<'a>),
+}
+
+impl<'a> Parse<'a> for HeapType<'a> {
+    fn parse(parser: Parser<'a>) -> Result<Self> {
+        let mut l = parser.lookahead1();
+        if l.peek::<Index>()? {
+            Ok(HeapType::Concrete(parser.parse()?))
+        } else if l.peek::<kw::shared>()? {
+            parser.parens(|p| {
+                p.parse::<kw::shared>()?;
+                Ok(HeapType::Abstract {
+                    shared: true,
+                    ty: p.parse()?,
+                })
+            })
```

Comment:
I haven't added `*.wast` tests for this PR because of the issue with checking shared-ness of concrete types. Until I can sort out how to properly check if a type is shared, I can't validate any of this. I did mention below that I wanted to try that in a separate PR, but I guess we should decide to either:
- disable `*.wast` parsing and accept this PR without tests
- fix the shared check and include the `*.wast` tests in this PR.